tools/shield_utils.py

```diff
@@ -0,0 +1,427 @@
+                               online_numa_nodes))) > 0
+
+
+def _get_cpulist(numa_node_id, raw=False):
+    """Get the cpulist of a NUMA node.
+
+    Args:
+        * numa_node_id (int).
+        * raw (boolean): Default is False. If set to True, will return
+                         the content found in the cpulist information
+                         present in sysfs.
+     Returns:
+        * str: the content found in the NUMA node cpulist file.
+        * list of ints: ints corresponding to elements of a LIST FORMAT, as
+                        described into `cpuset` documentation.
+    """
+    cpulist_path = SYSFS_MOUNTPOINT + "/devices/system/node/node{}/cpulist"\
+        .format(numa_node_id)
+    with open(cpulist_path) as fp:
+        if raw:
+            return fp.readline().strip()
+        return _parse_list_format(fp.readline())
+
+
+def _is_memory_node_available(numa_node_id):
+    """Check if the memory node is used by an explicit cpuset.
+
+    By explicit cpusets we mean cpusets different than root and '/docker'
+
+    Args:
+        * numa_node_id (int).
+    Returns:
+        * boolean.
+    """
+    (root, dirnames, _) = next(os.walk(CPUSET_MOUNTPOINT))
+    for dirname in dirnames:
+        if dirname == "docker":
+            continue
+
+        with open(root + "/" + dirname + "/cpuset.mems") as fp:
+            log.say(root + "/" + dirname + "/cpuset.mems")
+            mems = _parse_list_format(fp.readline())
+            if not len(list(filter(lambda m: m == numa_node_id, mems))) == 0:
+                return False
+
+    return True
+
+
+def _get_non_available_cpus():
+    """Return a set non available CPU cores.
+
+    A non available CPU core is a cpu id which is used by
+    others cpuset than `/` (root) and `/docker`.
+
+    Returns:
+        * list of ints: a list of CPU core IDs used by other cpusets than `/`
+                        (root) and `/docker`.
+    """
+    (root, dirnames, _) = next(os.walk(CPUSET_MOUNTPOINT))
+    non_avail_cpus = set()
+    for dirname in dirnames:
+        if dirname == "docker":
+            continue
+
+        with open(root + "/" + dirname + "/cpuset.cpus") as fp:
+            non_avail_cpus |= set(_parse_list_format(fp.readline()))
+
+    return list(non_avail_cpus)
+
+
+def _get_available_cpus(numa_node_id):
+    """Determine NUMA node CPUs which are not assigned to an explicit cpuset.
+
+    By explicit cpusets we mean cpusets different than root and '/docker'.
+
+    Args:
+        * numa_node_id (int).
+    Returns:
+        * list of ints: a list of CPU core IDs free for use by docker
+                        containers.
+    """
+    cpulist = _get_cpulist(numa_node_id)
+    non_avail_cpus = _get_non_available_cpus()
+    avail_cpus = []
+
+    # For every NUMA node core, check if it is available. Add it to the
+    # avail_cpus list if it is not present inside the non_avail_cpus list.
+    for cpu in cpulist:
+        cpu_is_avail = True
```

Comment:
Maybe there's something like `iter.filter()` in rust.

Reply:
Done.